homeassistant/components/matrix/__init__.py

```diff
@@ -49,10 +52,10 @@
     {
         DOMAIN: vol.Schema(
             {
-                vol.Required(CONF_HOMESERVER): cv.url,
+                vol.Optional(CONF_HOMESERVER): cv.url,
                 vol.Optional(CONF_VERIFY_SSL, default=True): cv.boolean,
-                vol.Required(CONF_USERNAME): cv.matches_regex("@[^:]*:.*"),
-                vol.Required(CONF_PASSWORD): cv.string,
+                vol.Optional(CONF_USERNAME): cv.matches_regex("@[^:]*:.*"),
```

Comment:
This is because if we use the UI to set up matrix HOMESERVER, USERNAME and PASSWORD are not needed. If they are still required it will not work with a mix of webui setting and yaml (for notification and rooms setting)

Reply:
The config schema here is only for the configuration yaml. That data will be imported to the import step, so will not pass via the user data form in the config flow.

Reply:
If I roll back to Required for these parameters I have this error when they are not set : 
```
2021-03-15 15:13:36 ERROR (MainThread) [homeassistant.config] Invalid config for [matrix]: required key not provided @ data['matrix']['homeserver']. Got None
required key not provided @ data['matrix']['password']. Got None
required key not provided @ data['matrix']['username']. Got None. (See /home/bpoirriez/Sources/home-assistant/core/config/configuration.yaml, line 27). Please check the docs at https://www.home-assistant.io/integrations/matrix
2021-03-15 15:13:36 ERROR (MainThread) [homeassistant.setup] Setup failed for matrix: Invalid config
```

Reply:
Yes, they are still required in the configuration yaml, if that is provided. The config flow validation and configuration yaml validation are two separate things.

Reply:
Yes but the config yaml is still needed to configure rooms and commands since list and map can not be set with config flow. So the matrix configuration will be an hybrid between config flow and yaml

Reply:
That's not the idea with the import. The idea is to import the configuration yaml and then deprecate it.

I think we need to think about how this integration should be migrated to our new standards. 

Maybe we can make an options flow to add one dict with word or expression match criteria at a time? Can we get the list of rooms the user has joined? Then we could allow the user to choose what rooms to include in the matching criteria.

Reply:
The starting point for this PR was this review https://github.com/home-assistant/core/pull/37625#discussion_r469843869
In the following comments I have pointed the difficulty to integrate the configuration with the config flow. If it is not possible with the current config flow support to have some list and dict configuration ans if a mix of config flow and yaml file is not acceptable, then I in favor of dropping this PR. In this case, I hope that the linked one (images support) can be merged even if there's a modification un the yaml and no config flow support.

Reply:
My suggestion is to build an options flow without trying use exactly the same approach as in the current configuration yaml, but still enable the same features.

Reply:
I agree with @bestlibre, I don't see how building a whole new options flow has any bearing on whether or not the integration can send images in notifications...